src/Stryker.Core/Stryker.Core/Mutants/MutationContext.cs

```diff
@@ -30,59 +22,76 @@ public MutationContext(CsharpMutantOrchestrator mutantOrchestrator)
         {
-            _ancestor = parent;
             _mainOrchestrator = parent._mainOrchestrator;
             InStaticValue = parent.InStaticValue;
+            Store = parent.Store;
+            FilteredMutators = parent.FilteredMutators;
+            FilterComment = parent.FilterComment;
         }
 
+        public IEnumerable<Mutant> GenerateMutantsForNode(SyntaxNode node) => _mainOrchestrator.GenerateMutationsForNode(node, this);
+
+        public INodeMutator FindHandler(SyntaxNode node) => _mainOrchestrator.GetHandler(node);
+
         /// <summary>
         ///  True when inside a static initializer, fields or accessor.
         /// </summary>
         public bool InStaticValue { get; set; }
 
         public bool MustInjectCoverageLogic => _mainOrchestrator.MustInjectCoverageLogic;
 
-        public bool HasBlockLevelMutant => BlockLevelControlledMutations.Count > 0;
-        
-        public bool HasStatementLevelMutant => StatementLevelControlledMutations.Count > 0 || HasBlockLevelMutant;
+        public Mutator[] FilteredMutators { get; private set; }
 
-        public MutationContext EnterStatic()
-        {
-            return new MutationContext(this) {InStaticValue = true};
-        }
+        public string FilterComment { get; set; }
 
-        public MutationContext Clone()
+        public MutationContext EnterStatic() => new(this) {InStaticValue = true};
+
+        public MutationContext Enter(MutationControl control)
         {
-            return new MutationContext(this);
+            switch (control)
+            {
+                case MutationControl.Statement:
+                    Store.EnterStatement();
+                    return this;
+                case MutationControl.Block:
+                    Store.EnterBlock();
+                    return new MutationContext(this);
+            }
+
+            return this;
         }
 
-        public void Discard()
+        public MutationContext FilterMutators(bool mode, Mutator[] filteredMutators, bool newContext,
+            string comment)
         {
-            if (HasStatementLevelMutant)
+            var result = newContext ? new MutationContext(this) : this;
+            if (mode)
             {
-                // some mutants 
-                _logger.LogDebug($"{BlockLevelControlledMutations.Count+StatementLevelControlledMutations.Count} mutations were not injected.");
-                foreach (var mutant in BlockLevelControlledMutations.Union(StatementLevelControlledMutations))
-                {
-                    mutant.ResultStatus = MutantStatus.CompileError;
-                    mutant.ResultStatusReason = "Stryker was not able to inject mutation in code.";
-                }
-                BlockLevelControlledMutations.Clear();
-                StatementLevelControlledMutations.Clear();
+                result.FilteredMutators = filteredMutators;
+            }
+            else if (result.FilteredMutators != null)
```

Comment:
```suggestion
            else if (result.FilteredMutators is not null)
```

All these fancy new csharp9 features but no is not null pattern? ;)